### PR TITLE
1977 annotation skip invalid images

### PR DESCRIPTION
**Pull request by Jarvi-Izana**
It depends on #2227, which will make `unity_sframe::append` work as the same as what frontend part does.

The new behavior will:
1. prohibit undefined images from being shown in the annotation GUI window.
2. will throw if no valid images there to be ready to render in GUI.
3. will return result-set including all rows having missing values in their `image` column, with the corresponding annotation column value being set as None (undefined).
```
In [6]: images = tc.SFrame("my_image/")
In [6]: images = tc.SFrame("my_image/")

In [7]: images
Out[7]:
Columns:
        path    str
        image   Image

Rows: 2

Data:
+-------------------------------+------------------------+
|              path             |         image          |
+-------------------------------+------------------------+
| /Users/guihaoliang/Work/te... | Height: 412 Width: 430 |
|              None             |          None          |
+-------------------------------+------------------------+
[2 rows x 2 columns]

In [8]: tc.image_classifier.annotate(images)
Out[8]:
Columns:
        path    str
        image   Image
        annotations     str

Rows: 2

Data:
+-------------------------------+------------------------+-------------+
|              path             |         image          | annotations |
+-------------------------------+------------------------+-------------+
| /Users/guihaoliang/Work/te... | Height: 412 Width: 430 |     Cat     |
|              None             |          None          |     None    |
+-------------------------------+------------------------+-------------+
[2 rows x 3 columns]

In [9]: tc.image_classifier.annotate(images[1:2])
---------------------------------------------------------------------------
ToolkitError                              Traceback (most recent call last)
<ipython-input-11-05069e9b96e8> in <module>
----> 1 tc.image_classifier.annotate(images[1:2])

~/Work/guicreate-1/debug/src/python/turicreate/toolkits/image_classifier/_annotate.py in annotate(data, image_column, annotation_column)
    111                             data,
    112                             [image_column],
--> 113                             annotation_column
    114                         )
    115     if __platform == "linux" or __platform == "linux2":

~/Work/guicreate-1/debug/src/python/turicreate/extensions.py in <lambda>(*args, **kwargs)
    168
    169 def _make_injected_function(fn, arguments):
--> 170     return lambda *args, **kwargs: _run_toolkit_function(fn, arguments, args, kwargs)
    171
    172 def _class_instance_from_name(class_name, *arg, **kwarg):

~/Work/guicreate-1/debug/src/python/turicreate/extensions.py in _run_toolkit_function(fnname, arguments, args, kwargs)
    157     if not ret[0]:
    158         if len(ret[1]) > 0:
--> 159             raise _ToolkitError(ret[1])
    160         else:
    161             raise _ToolkitError("Toolkit failed with unknown error")

ToolkitError: Expect positive number of images. Found 0 valid images and 1 missing images.
```

### CODE REVIEW

**Comment by Jarvi-Izana** [comment]
Big thanks to @abhishekpratapa for explaining the annotation code design.

File: src/visualization/annotation/annotation_base.cpp [hunk 1]
```diff
@@ -95,20 +117,18 @@ std::shared_ptr<annotation_global> AnnotationBase::get_annotation_registry() {
...
+  DASSERT_TRUE(m_data->contains_column("__idx"));
+  const auto &image_column_name = m_data_columns.at(0);
+  auto split_data_set =
+      m_data->drop_missing_values({image_column_name}, how, true, recursive);
+  m_data = std::static_pointer_cast<unity_sframe>(split_data_set.front());
+  m_data_na = std::static_pointer_cast<unity_sframe>(split_data_set.back());
+}
 
-  m_data->add_column(empty_annotation_sarray, "__idx");
+void AnnotationBase::_addIndexColumn() {
```
Comment (Jarvi-Izana) [inline src/visualization/annotation/annotation_base.cpp]: more efficient impl of constructing seq.

File: src/python/turicreate/data_structures/sframe.py [hunk 2]
```diff
@@ -3778,48 +3778,8 @@ def append(self, other):
         if type(other) is not SFrame:
             raise RuntimeError("SFrame append can only work with SFrame")
```
Comment (Jarvi-Izana) [inline src/python/turicreate/data_structures/sframe.py]: code change for #2227; irrelevant to this PR; Feel free to ignore.

File: src/core/storage/sframe_interface/unity_sframe.cpp [hunk 3]
```diff
@@ -426,9 +426,9 @@ std::shared_ptr<unity_sarray_base> unity_sframe::select_column(const std::string
   size_t column_index = _column_index_iter - _column_names.begin();
   auto ret = select_column(column_index);
```
Comment (Jarvi-Izana) [inline src/core/storage/sframe_interface/unity_sframe.cpp]: code change for #2227; irrelevant to this PR; Feel free to ignore.

File: src/visualization/annotation/annotation_base.cpp [hunk 4]
```diff
@@ -58,11 +58,32 @@ void AnnotationBase::_sendProgress(double value) {
...
+      if (cast_type != na_type) {
+        auto data_sarray = std::static_pointer_cast<unity_sarray>(
+            m_data_na->select_column(m_annotation_column));
+        auto integer_annotations =
+            data_sarray->astype(cast_type, true);
+        m_data_na->remove_column(m_data_na->column_index(m_annotation_column));
+        m_data_na->add_column(integer_annotations, m_annotation_column);
+      }
+      copy_data =
+          std::static_pointer_cast<unity_sframe>(m_data->append(m_data_na));
```
Comment (Jarvi-Izana) [inline src/visualization/annotation/annotation_base.cpp]: `append` requires the change #2227 to behave correctly.

File: src/core/storage/sframe_interface/unity_sframe.cpp (hunk 3 above)
Comment (znation) [reply src/core/storage/sframe_interface/unity_sframe.cpp]: Can you rebase your changes against master? It's better (for reviewers) if you put up independent PRs. Thanks!

Comment (znation) [reply src/core/storage/sframe_interface/unity_sframe.cpp]: Nevermind, I see you actually have a dependency here.

File: test/annotation/image_classifier_tests.cxx [hunk 5]
```diff
@@ -346,7 +346,7 @@ struct image_classification_test {
 
     TS_ASSERT(annotation_testing::check_equality(annotation_sf, returned_sf));
 
-    
+
```
Comment (znation) [inline test/annotation/image_classifier_tests.cxx]: Since this is only a whitespace change, please revert the changes in this file.

File: userguide/image_classifier/advanced-usage.md [hunk 6]
```diff
@@ -20,7 +20,12 @@ annotated_data = tc.image_classifier.annotate(data)
 
 ```
 
-If you forget to assign the output of your annotation to a variable, 
+This utility will **only collect valid images**. All undefined values
```
Comment (znation) [inline userguide/image_classifier/advanced-usage.md]: I'm not sure what "only collect valid images" means.

How about "only present rows from the SFrame that have non-missing images"? And I don't think I would make it bold... as a user, I would not think it's too surprising or needs to be taken note of, just good to know.

File: src/visualization/annotation/annotation_base.cpp [hunk 7]
```diff
@@ -58,11 +58,32 @@ void AnnotationBase::_sendProgress(double value) {
...
+    } else {
+      copy_data = std::static_pointer_cast<unity_sframe>(
           m_data->copy_range(0, 1, m_data->size()));
+    }
+  }
+  // temporary solution
```
Comment (znation) [inline src/visualization/annotation/annotation_base.cpp]: What is temporary about this solution? What problem is it solving? Why is the solution temporary?

Or if this comment is outdated, please remove it.

File: src/visualization/annotation/annotation_base.cpp [hunk 8]
```diff
@@ -58,11 +58,32 @@ void AnnotationBase::_sendProgress(double value) {
...
+  std::shared_ptr<unity_sframe> copy_data;
+  {
+    // append undefined values
+    if (m_data_na && m_data_na->size()) {
+      auto cast_type = m_data->select_column(m_annotation_column)->dtype();
+      auto na_type = m_data_na->select_column(m_annotation_column)->dtype();
+      if (cast_type != na_type) {
+        auto data_sarray = std::static_pointer_cast<unity_sarray>(
+            m_data_na->select_column(m_annotation_column));
+        auto integer_annotations =
```
Comment (znation) [inline src/visualization/annotation/annotation_base.cpp]: These aren't necessarily integer annotations, right? Couldn't they be string annotations (depending on cast_type)?

File: src/visualization/annotation/annotation_base.cpp [hunk 9]
```diff
@@ -73,18 +94,19 @@ AnnotationBase::returnAnnotations(bool drop_null) {
...
     annotation_global->annotation_sframe = copy_data;
     return copy_data;
-  }
+    }
 
-  std::vector<std::string> annotation_column_name = {m_annotation_column};
-  std::list<std::shared_ptr<unity_sframe_base>> dropped_missing =
-      copy_data->drop_missing_values(annotation_column_name, false, false, false);
+    std::vector<std::string> annotation_column_name = {m_annotation_column};
+    std::list<std::shared_ptr<unity_sframe_base>> dropped_missing =
```
Comment (znation) [inline src/visualization/annotation/annotation_base.cpp]: I'm not sure I understand the overall approach here. Why do we drop missing values at all, only to add them back in upon return? Wouldn't it be simpler to send all rows (including those with missing data) and receive all rows? We don't necessarily need to display those rows on the client, in order to send and receive them.

If this approach was considered and rejected, please add a comment here as to why; if not, I think we should consider it.

File: src/visualization/annotation/annotation_base.hpp [hunk 10]
```diff
@@ -90,6 +90,9 @@ class EXPORT AnnotationBase : public ml_model_base {
 
   virtual void checkDataSet() = 0;
 
+  /* must be called after construction */
```
Comment (znation) [inline src/visualization/annotation/annotation_base.hpp]: If it's intended to be called once, and only from the constructor, I wouldn't put it in the public API. It should ideally be private in that case and non-virtual (or protected and virtual, if subclass constructors need to call it directly instead of through their superclass constructor).

File: test/annotation/image_classifier_tests.cxx (hunk 5 above)
Comment (Jarvi-Izana) [reply test/annotation/image_classifier_tests.cxx]: Yeah, I could. IMO, it can keep the source pkg smaller XD. Besides, you can set ignore whitespaces change when you do code review. 
https://github.blog/2018-05-01-ignore-white-space-in-code-review/
@znation

File: src/visualization/annotation/annotation_base.cpp (hunk 8 above)
Comment (Jarvi-Izana) [reply src/visualization/annotation/annotation_base.cpp]: Yeah, but the first line `cast` will try to probe whether the annotation can be convertible to integer or not. This is just to align with it.

File: src/visualization/annotation/annotation_base.cpp (hunk 7 above)
Comment (Jarvi-Izana) [reply src/visualization/annotation/annotation_base.cpp]: Haha, I forget to change its position.

Well, the temp solution is to sort by `__idx` column to restore the original order, which is not that efficient. But in the future, if we hit any perf issue, we should come up with a solution that we don't have to split the image set into defined part and undefined part. We can do all operations inline. Since the design by @abhishekpratapa requires all images feeding into neural network to be valid, we come up with this **temp** solution.

File: userguide/image_classifier/advanced-usage.md (hunk 6 above)
Comment (Jarvi-Izana) [reply userguide/image_classifier/advanced-usage.md]: Oh, sure.

File: src/visualization/annotation/annotation_base.hpp (hunk 10 above)
Comment (Jarvi-Izana) [reply src/visualization/annotation/annotation_base.hpp]: That's a good point! I should provide it as a private virtual.

File: src/visualization/annotation/annotation_base.cpp (hunk 9 above)
Comment (Jarvi-Izana) [reply src/visualization/annotation/annotation_base.cpp]: Sure. It's explained just now in previous threads.

In short, we can do everything inline but it's tricky and requires non-trivial work. Maybe @abhishekpratapa can explain it more. So we designed this workaround with the consideration that annotation is done by humans and each batch job won't be that large (less than 1000 pics per batch) for them. So split the data set won't be the bottle neck.

File: src/visualization/annotation/annotation_base.hpp (hunk 10 above)
Comment (Jarvi-Izana) [reply src/visualization/annotation/annotation_base.hpp]: Ah, I just realized it's a mistake to call virtual function during construction. I will make it non-virtual protected.

Comment (Jarvi-Izana) [reply src/visualization/annotation/annotation_base.hpp]: it's tricky to have virtual functions involved in construction. I will make it non-virtual for this case.

Comment (Jarvi-Izana) [reply src/visualization/annotation/annotation_base.hpp]: @znation as is recorded in #2234. Will be a feature request for @abhishekpratapa. For now, it's okay since base class is pure virtual and derived class directly inherits base. If the inheritance chain grows, then we should consider changing the code structure first. For now, it's irrelevant to this pr.

**Comment by Jarvi-Izana** [comment]
referenced by #1977

**Comment by znation** [comment]
Please rebase against latest master; this PR now has changes in it that (seem to be) already in master.

**Comment by Jarvi-Izana** [comment]
@znation rebased.

**Comment by Jarvi-Izana** [comment]
Once it passes the internal CI, I will merge it.
